src/dynamic.ts

```diff
@@ -0,0 +1,71 @@
+import { EventEmitter } from 'events'
+import type { FixedThreadPoolOptions, WorkerWithMessageChannel } from './fixed'
+import FixedThreadPool from './fixed'
+
+class MyEmitter extends EventEmitter {}
+
+export type DynamicThreadPoolOptions = FixedThreadPoolOptions
+
+/**
+ * A thread pool with a min/max number of threads, is possible to execute tasks in sync or async mode as you prefer.
+ *
+ * This thread pool will create new workers when the other ones are busy, until the max number of threads,
+ * when the max number of threads is reached, an event will be emitted, if you want to listen this event use the emitter method.
+ *
+ * @author [Alessandro Pio Ardizio](https://github.com/pioardi)
+ * @since 0.0.1
+ */
+export default class DynamicThreadPool<
+  Data = any,
+  Response = any
+> extends FixedThreadPool<Data, Response> {
+  public readonly emitter: MyEmitter
+
+  /**
+   * @param min Min number of threads that will be always active
+   * @param max Max number of threads that will be active
+   * @param filename A file path with implementation of `ThreadWorker` class, relative path is fine.
+   * @param opts An object with possible options for example `errorHandler`, `onlineHandler`. Default: `{ maxTasks: 1000 }`
+   */
+  public constructor (
+    public readonly min: number,
+    public readonly max: number,
+    public readonly filename: string,
+    public readonly opts: DynamicThreadPoolOptions = { maxTasks: 1000 }
+  ) {
+    super(min, filename, opts)
+
+    this.emitter = new MyEmitter()
+  }
+
+  protected _chooseWorker (): WorkerWithMessageChannel {
+    let worker: WorkerWithMessageChannel
+    for (const entry of this.tasks) {
+      if (entry[1] === 0) {
+        worker = entry[0]
+        break
+      }
+    }
+
+    if (worker) {
+      // a worker is free, use it
+      return worker
+    } else {
+      if (this.workers.length === this.max) {
+        this.emitter.emit('FullPool')
+        return super._chooseWorker()
+      }
+      // all workers are busy create a new worker
+      const worker = this._newWorker()
+      worker.port2.on('message', (message) => {
+        if (message.kill) {
+          worker.postMessage({ kill: 1 })
+          worker.terminate()
+        }
+      })
+      return worker
+    }
+  }
+}
+
+module.exports = DynamicThreadPool
```

Comment:
These `module.exports` necessary to not break compatibility for now
Later we should remove this together with `"esModuleInterop": true` in `tsconfig.json` and introduce a breaking change release

Reply:
Are you sure about that? 
The transpilation of export default class should build javascript module export compatible on its own given the right options.

Reply:
Yes I'm sure and tried it locally  
If not using this (_legacy_) syntax, it is a breaking change
IMO we should make clean PRs/commits instead of just shift everything at once. In this way it's more clear what happend and it's more incrementally

Reply:
There's still something that I do not understand: 
+ You have remove `module.exports.ThreadWorker = ThreadWorker` inside its class file;
+ You have kept `module.exports = DynamicThreadPool` and `module.exports = FixedThreadPool` inside their respective class files;
+ And there's `module.exports = { FixedThreadPool, DynamicThreadPool, ThreadWorker }` inside the index.ts file. 

I think the legacy module export syntax is only needed in the index.ts file and can be removed elsewhere.